날씨앱 [STEP 1] roks,bello

KimRoks opened the pull request:
리뷰어: @protocorn93

버드: @JinUng41, @KimRoks

안녕하세요, 콘!!

은행창구 관리 앱으로 만나뵙게 된 벨로, 록스입니다.

[STEP 1]에서는 API 통신을 통해 받는 JSON의 형식을 고려하여 모델 타입을 구현하고

URL Session을 활용하여 네트워킹 작업을 수행하였습니다!

이번 작업을 진행하며 Style Share의 Commit Convension을 일부 착안하여 
한줄에 100자가 넘지 않도록 작업해보았습니다.
https://github.com/StyleShare/swift-style-guide

### 구현 내용

- [x]  파싱한 `JSON` 데이터와 매핑할 모델 설계
- [x]  `CodingKeys` 프로토콜 활용
- [x]  `API` 서버와 통신할 인스턴스 역할을 수행할 타입 구현
- [x]  범용성과 재사용성, 유연성을 고려한 네트워킹 타입 구현
- [x]  라이브러리 없이 `URLSession`을 활용하여 구현

## 주요 구현 내용

### 디렉터리 구조

```
.
├── WeatherAPI.plist
├── WeatherForecast
│   ├── Controllers
│   │   └── ViewController.swift
│   ├── Extensions
│   │   └── Bundle.swift
│   ├── Info.plist
│   ├── Models
│   │   ├── ForecastType.swift
│   │   ├── NetworkError.swift
│   │   ├── NetworkManager.swift
│   │   └── Weather
│   │       ├── CurrentWeather.swift
│   │       └── ForecastWeather.swift
│   ├── Resources
│   │   ├── AppDelegate.swift
│   │   ├── Assets.xcassets
│   │   │   ├── AccentColor.colorset
│   │   │   │   └── Contents.json
│   │   │   ├── AppIcon.appiconset
│   │   │   │   └── Contents.json
│   │   │   └── Contents.json
│   │   └── SceneDelegate.swift
│   └── Views
│       └── Base.lproj
│           ├── LaunchScreen.storyboard
│           └── Main.storyboard
```

### API_KEY 숨기는 방법

API통신을 진행하며 보안 상 API KEY를 숨기는것이 좋겠다고 판단하였습니다

저희가 고려한 방법들은 아래와 같습니다

1. **타입 인스턴스 활용**
    - 방법: 어디서든 사용할 수 있게끔 API 키를 `static`으로 선언하고 사용하는 방식
    - 장점: 편리하게 어디서든 사용 가능
    - 단점: 커밋할 때마다 수동으로 키를 관리해야 하는 번거로움이 있어 선택하지 않았습니다.
2. **xcconfig 활용**
    - **방법**:XCConfig 파일에 API 키를 작성하고 Build Setting 및 Gitignore를 활용하여 사용
    - 장점: 빌드 설정을 통한 유연한 관리 가능
    - 단점: 복잡성이 높아 프로세스가 번거로울 수 있어 최종적으로 선택하지 않았습니다.
3. **info.plist ✅**
    - **방법**:Gitignore에 API Key를 포함한 Info.plist를 추가하고, Bundle을 Extension하여 사용
    - **장점**: 간단하고 명확한 방법, Gitignore를 통한 보안 강화
    - **선택 이유**: 본 방법이 비교적 간편하면서도 보안 측면에서 충분한 수준의 안정성을 제공한다고 판단해서 최종적으로 선택하게 되었습니다.

### 현재 날씨와 5일 예보의 공통되는 부분들에 대한 처리

- 현재 날씨의 JSON 예시
    
    [https://openweathermap.org/current#:~:text=in API response.-,JSON,-JSON format API](https://openweathermap.org/current#:~:text=in%20API%20response.-,JSON,-JSON%20format%20API)
    
- 5일 날씨 예보 JSON 예시
    
    [https://openweathermap.org/forecast5#:~:text=in API response.-,JSON,-JSON format API](https://openweathermap.org/forecast5#:~:text=in%20API%20response.-,JSON,-JSON%20format%20API)
    
- 위의 두 링크에 나와있는 JSON 예시를 참고하여 모델을 구현해 보았습니다.
- 프로젝트 코드에서는 CurrentWeather, ForecastWeather 클래스에서 사용하는 객체가 겹치기 때문에 겹치는 객체의 선언은 CurrentWeather.swift 내에서 선언하였습니다.

### 네트워킹 에러처리

- NetworkError 타입을 구현하여 여러 가지의 상황에 대처할 수 있도록 구현하였습니다.
    - dataTask 메서드를 통해 error가 발생하면 발생한 해당 error를 출력합니다.
    - response에서 응답코드가 2xx대가 아니라면 해당 응답 코드를 출력합니다.
    - data가 디코딩이 되지 않았을 때 실패 메세지를 출력합니다.

### 들어오는 데이터 확인

- 다음의 코드를 통해 들어오는 데이터를 확인할 수 있었습니다.
    
    ```swift
    NetworkManager().fetchWeather(weatherType: .current, latitude: 37.820072649485134, longitude: -122.47823364517218) { (result: Result<CurrentWeather, Error>) in
        switch result {
        case .success(let weather):
            print(weather)
        case .failure(let error):
            print(error)
        }
    }
    ```
    

## 궁금한 점

### CurrentWeather (현재 날씨) 와 ForecastWeather (5일 예보) 모두 사용되는 객체의 정의

- 두 날씨에 대해서 모두 사용되는 객체의 정의를 현재는 CurrentWeather.swift에 정의하였습니다.
- 그리고 ForecastWeather.swift에서는 따로 필요한 객체에 대해서만 정의 하였습니다.
- 만일 다른 방식으로 구현해 본다면 다음과 같이 구현할 수도 있겠다라는 생각이 들었습니다.
    1. 공통으로 사용되는 객체를 따로 파일을 만들어서 해당 파일에 정의
    2. 중첩 타입으로 정의하여 두 날씨에 대해 각각 정의
- 따라서 저희가 구현한 방식에 대해서 조언을 듣고 싶습니다!!

### URLSession의 싱글톤 인스턴스에 관하여

- URLSession을 구현하며 URLSession은 기본적으로 싱글톤으로 shared Session을 제공하고 있습니다.
- 해당 객체에서의 작업들은 기본적으로 캐시, 쿠키 저장소 및 인증 내용을 공유한다고 학습하였습니다
- 또한 필요한 것 이상으로 세션을 만들지말라는 내용도 학습하였습니다.
- 최종적으론 shared내에서 두번의 네트워킹 작업을 수행하였고 이에 관해 궁금증이 생겨 의견을 듣고싶습니다

1. shared 세션 외의 Session은 단순히 작업분류 용도가  아닌 꼭 필요한 경우에만 사용하는것인가요 ? 
    
    ex) delegate 정의, 다운로드 작업
    

 2.  캐시나 쿠키를 공유하는것이 불필요하다고 느껴지는데 이는 크게 신경써야 할 부분이 아닌것일까요?
      신경을 써야한다면 어떻게 처리해야할지도 궁금합니다.

JinUng41 committed:
config: gitignore 설정

JinUng41 committed:
config: minimum target 14.0으로 수정

KimRoks committed:
config: 파일 폴더링

KimRoks committed:
feat: Weather 모델 정의

- Current Weather 정의

KimRoks committed:
feat: FiveDaysWeather 모델 정의

JinUng41 committed:
refactor: 폴더 구조 변경 및 NetworkManager 생성

JinUng41 committed:
feat: API_KEY 숨김 처리 및 Bundle 익스텐션 정의

KimRoks committed:
test: git ignore test

JinUng41 committed:
refactor: 5일 기상 예보 구조체 이름 변경

KimRoks committed:
chore: 마크 주석 네이밍 수정

KimRoks committed:
chore: 주석 수정

JinUng41 committed:
refactor: 중복되는 내용을 하나로 합침.

KimRoks committed:
feat: NetwokrManager 정의, WeatherURL Enum 정의

JinUng41 committed:
refactor: 관측 정보에 따른 URL을 리턴하도록 구현

JinUng41 committed:
chore: 주석 수정

JinUng41 committed:
refactor: CurrentWeather에 대해서 일부를 옵셔널 타입으로 구현

JinUng41 committed:
refactor: fetchCurrentWeather 수정

URL을 가져오는 방식을 변경

KimRoks committed:
feat: forecast NetworkManager 정의

JinUng41 committed:
refactor: 현재 날씨의 모델 내부 변수 이름 수정

JinUng41 committed:
refactor: 5일 예보 객체 구현

KimRoks committed:
refactor: Bundle init 변경

- init(contentsOfFile:)가 Deprecated 된 이유로 init(contentsOf:)로 변경

JinUng41 committed:
refactor: 테스트 코드 삭제

JinUng41 committed:
feat: 제네릭 타입으로 날씨를 받는 함수를 구현

날씨를 받는 함수가 제네릭으로 구현되어 어떠한 종류가 와도 대체가 가능함.

KimRoks committed:
refactor: 모델 네이밍 변경

KimRoks committed:
feat: Network Error case 정의 및 적용

JinUng41 committed:
refactor: 응답 에러에서 응답 코드를 같이 실어서 출력할 수 있도록 수정

KimRoks committed:
refactor: API 매개변수 추가

- 온도 표시를 화씨 -> 섭씨

JinUng41 committed:
refactor: NetworkError의 unknownError를 수정

dataTask 메서드를 통해 error가 있으면, error를 연관값으로 저장하여 출력하게끔 동작을 수정

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/ViewController.swift:
```diff
@@ -7,12 +7,8 @@
 import UIKit
 
 class ViewController: UIViewController {
```
상속이 계획된 것이 아니라면 `final`을 붙여주시면 좋을 것 같아요.

KimRoks replied:
확인했습니다!
불필요한 성능 저하를 막기위해 한번 더 검토하도록 하겠습니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Models/ForecastType.swift:
```diff
@@ -0,0 +1,22 @@
+    case current = "weather"
+    case forecast = "forecast"
+    
+    func fetchURL(lon: Double, lat: Double) -> String {
```
동사(fetch)가 앞으로 오기 때문에 행위를 의미해서 실질적인 데이터를 받아오는 행위로 오해할 수 있을 것 같아요. 명시적으로 makeURL 같이 만들어 반환한다는 의미가 담기면 좋을 것 같아요.

KimRoks replied:
명확한 의미전달을 할 수 있도록 네이밍 하도록 하겠습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/ForecastType.swift:
```diff
@@ -0,0 +1,22 @@
+    func fetchURL(lon: Double, lat: Double) -> String {
+        guard let apiKey = Bundle.main.apiKey else {
+            print("API KEY를 찾을 수 없습니다!!")
+            return ""
```
nil 과 빈 문자열 중 빈 문자열을 반환하기로 결정한 이유가 있을까요?

KimRoks replied:
사실 논의가 있었던 부분인데 깜빡하고 신경쓰지 못하고 넘어간 것 같습니다

반환 타입을 String에서 URL? 로 변경, 실패 시 nil을 반환하고
호출단에서 한번의 옵서녈 바인딩으로 사용할 수 있도록 변경하였습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/NetworkManager.swift:
```diff
@@ -0,0 +1,50 @@
+        
+        guard let url = URL(string: weatherType.fetchURL(lon: longitude, lat: latitude))
+        else {
+            print("올바른 URL이 아닙니다.")
```
여기선 에러와 함께 completion을 호출해주어야 하지 않을가요?

KimRoks replied:
추가적인 error case 정의와 함께 completion을 호출하도록 수정하였습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/NetworkManager.swift:
```diff
@@ -0,0 +1,50 @@
+            }
+            
+            guard let data = data else {
+                print("Error: Did not receive data")
```
여기서도 completion을 호출해주어야 하지 않을까요?

KimRoks replied:
이 또한 추가적인 error case 정의와 함께 completion을 호출하도록 수정하였습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/NetworkManager.swift:
```diff
@@ -0,0 +1,50 @@
+
+final class NetworkManager {
+    
+    func fetchWeather<T: Decodable>(weatherType: ForecastType,
```
`Decodable`한 제네릭 타입을 결과로 전달한다는 것을 잘 표현해주신 것 같아요.

하지만 조금 애매모호한 것 같아요. 좀 더 추상화될 수 있을 것 같아요. `NetworkManager`라면 다른 API 호출을 위해서도 사용되어야 할 것 같은데요. 

온전히 네트워킹 호출의 역할만 가져가면 좋을 것 같아요. API 요청에 관한 정보들을 좀 더 추상화(`URLRequest`)하여 추상화된 정보만 갖고 단순 네트워킹 호출만 하는 객체가 되면 좋을 것 같아요.

JinUng41 replied:
`APIRequest`라는 프로토콜을 선언하고 해당 프로토콜을 따르는 `WeatherRequest`를 추가 선언하여 NetworkManager에서는 `APIRequest`를 따르는 어떠한 API 요청이라도 네트워크 작업을 진행하도록 구현하였습니다!!

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Weather/CurrentWeather.swift:
```diff
@@ -0,0 +1,98 @@
+
+// MARK: - CurrentWeather
+
+struct CurrentWeather: Codable {
```
`CurrentWeather`는 `Encoding`도 될 수 있는 것일까요? `Decoding`만 되는것이라면 `Codable`을 따를 필요는 없어보여요. 아래 DTO들도 마찬가지에요

KimRoks replied:
채택된 프로토콜만 보고도 역할을 알 수 있도록 
Encoding이 필요없는 작업들에 대해 전부 Decodable을 채택하도록 변경하였습니다

protocorn93 commented on WeatherForecast/WeatherForecast/Models/Weather/CurrentWeather.swift:
```diff
@@ -0,0 +1,98 @@
+
+// MARK: - Sys
+
+struct Sys: Codable {
+    let type, id: Int
+    let country: String
+    let sunrise, sunset: Int
+}
```
서버에서 내려오는 키 값에 모델명도 맞출 필요는 없지 않나 싶어요. Sys가 어떤 역할을 하는지 네이티브 코드단에선 표현해주면 좋을 것 같아요.

KimRoks replied:
서버에서 내려주는 sys 라는 이름과 역할이 너무 애매모호한거 같아 보류된 부분이었습니다
적절한 네이밍을 선정하여 반영하도록 하겠습니다

protocorn93 reviewed (changes requested):
두 분 깔끔한 코드로 잘 구현해주신 것 같아요. 코드 코멘트 이외에 몇 가지 의견 남겨 드려요. 

1. 커밋 로그를 깔끔하게 유지해주셨으면 좋겠어요. (git의 `fixup` 기능을 활용해주시면 도움이 될 것 같아요.)
2. 단위 테스트 코드 작성에 대해 고민해보시면 좋을 것 같아요. 현재 코드에선 어떤 객체를 테스트 해야할까?, 테스트 코드를 작성할 수 있을까?에 대해 고민해보시면 좋을 것 같아요.


## 질문

> CurrentWeather (현재 날씨) 와 ForecastWeather (5일 예보) 모두 사용되는 객체의 정의
>
> 두 날씨에 대해서 모두 사용되는 객체의 정의를 현재는 CurrentWeather.swift에 정의하였습니다.
그리고 ForecastWeather.swift에서는 따로 필요한 객체에 대해서만 정의 하였습니다.
만일 다른 방식으로 구현해 본다면 다음과 같이 구현할 수도 있겠다라는 생각이 들었습니다.
공통으로 사용되는 객체를 따로 파일을 만들어서 해당 파일에 정의
중첩 타입으로 정의하여 두 날씨에 대해 각각 정의
따라서 저희가 구현한 방식에 대해서 조언을 듣고 싶습니다!!

분명한 것은 이건 팀의 상황 회사의 상황 그리고 서버 엔지니어와의 커뮤니케이션에 따라 달라질 수 있어요. 다만 외부 API를 사용하는 것이기 때문에 상당수의 프로퍼티가 같을지라도 별도로 분리하는게 좋을 것 같아요. 언제 API가 어떻게 변경될지 보장되지 않기 때문이에요. 


> URLSession의 싱글톤 인스턴스에 관하여
> 
> URLSession을 구현하며 URLSession은 기본적으로 싱글톤으로 shared Session을 제공하고 있습니다.
해당 객체에서의 작업들은 기본적으로 캐시, 쿠키 저장소 및 인증 내용을 공유한다고 학습하였습니다
또한 필요한 것 이상으로 세션을 만들지말라는 내용도 학습하였습니다.
최종적으론 shared내에서 두번의 네트워킹 작업을 수행하였고 이에 관해 궁금증이 생겨 의견을 듣고싶습니다
shared 세션 외의 Session은 단순히 작업분류 용도가 아닌 꼭 필요한 경우에만 사용하는것인가요 ?

ex) delegate 정의, 다운로드 작업

캐시나 쿠키를 공유하는것이 불필요하다고 느껴지는데 이는 크게 신경써야 할 부분이 아닌것일까요?
신경을 써야한다면 어떻게 처리해야할지도 궁금합니다.

네네 현재는 별도의 세션을 별도로 구현할 필요는 없어보여요. 일반적인 케이스에선 별도로 별도의 Configuration을 만들어 세션을 만들 필요는 없는데요. 좀 더 정교한 핸들링이 필요한 상황에선 별도 구현이 필요하긴해요 (비디오 다운로드 등) 또한 캐시나 쿠키 관련해서도 현재 수준에선 크게 신경쓰실 부분은 없어보이네요.

KimRoks committed:
refactor: 즉각 처리 가능한 피드백 적용

- 상속이 필요없는 ViewController에 final 키워드 추가
- 명확한 표현을 위해 fetchURL() -> makeURL() 로 네이밍 변경
- 인코딩 작업이 필요없는 타입들의 프로토콜을 Codable -> Decodable로 변경

JinUng41 committed:
refactor: 추가적인 에러 케이스 정의 및 모든 경우에 에러로 대응하도록 수정

JinUng41 committed:
refactor: 모델 분리

- 현재 날씨와 5일 예보 모두 사용되는 객체를 각각으로 분리하고 네이밍을 수정
- 최상위 객체인 CurrentWeather와 ForecastWeather는 각각 Current와 Forecast로 수정

KimRoks committed:
refactor: CodingKey 재설정

- 가독성을 위해 모든 케이스에서 줄임말 삭제 조치

JinUng41 committed:
refactor: 종류에 상관없이 API 요청에 따른 네트워크 작업을 수행하도록 수정

- APIRequest 프로토콜을 선언
- 해당 프로토콜을 따르는 WeatherRequest를 추가적으로 선언
- WeatherType은 단순히 전달받을 날씨예보의 종류만을 정의
- NetworkError에서는 API_KEY가 없을 경우에 대한 케이스를 추가적으로 정의
- NetworkManager는 단순히 들어오는 API 요청에 대해서 네트워크 작업을 수행

JinUng41 committed:
refactor: 단위 테스트를 위한 NetworkManager 코드 변경

- URLSessionProtocol, URLSessionDataTaskProtocol를 정의
- 테스트 시에는 MockURLSession, MockURLSessionDataTask를 구현하고
- 실제 네트워크 통신은 하지 않는 유닛 테스트가 가능하도록 함
